src/backend/distributed/executor/insert_select_executor.c

```diff
@@ -0,0 +1,150 @@
+ * distributed table and then executing the SELECT query using that DestReceiver
+ * as the tuple destination.
+ */
+TupleTableSlot *
+CoordinatorInsertSelectExecScan(CustomScanState *node)
+{
+	CitusScanState *scanState = (CitusScanState *) node;
+	TupleTableSlot *resultSlot = NULL;
+
+	if (!scanState->finishedRemoteScan)
+	{
+		EState *executorState = scanState->customScanState.ss.ps.state;
+		MultiPlan *multiPlan = scanState->multiPlan;
+		Query *selectQuery = multiPlan->insertSelectQuery;
+		List *insertTargetList = multiPlan->insertTargetList;
+		Oid targetRelationId = multiPlan->targetRelationId;
+
+		ereport(DEBUG1, (errmsg("Collecting INSERT ... SELECT results on coordinator")));
+
+		ExecuteSelectIntoRelation(targetRelationId, insertTargetList, selectQuery,
+								  executorState);
+
+		scanState->finishedRemoteScan = true;
+	}
+
+	resultSlot = ReturnTupleFromTuplestore(scanState);
+
+	return resultSlot;
+}
+
+
+/*
+ * ExecuteSelectIntoRelation executes given SELECT query and inserts the
+ * results into the target relation, which is assumed to be a distributed
+ * table.
+ */
+static void
+ExecuteSelectIntoRelation(Oid targetRelationId, List *insertTargetList,
+						  Query *selectQuery, EState *executorState)
+{
+	ParamListInfo paramListInfo = executorState->es_param_list_info;
+
+	ListCell *insertTargetCell = NULL;
+	List *columnNameList = NIL;
+	bool stopOnFailure = false;
+
+	CitusCopyDestReceiver *copyDest = NULL;
+
+	BeginOrContinueCoordinatedTransaction();
+
+	/* build the list of column names for the COPY statement */
+	foreach(insertTargetCell, insertTargetList)
+	{
+		TargetEntry *insertTargetEntry = (TargetEntry *) lfirst(insertTargetCell);
+
+		columnNameList = lappend(columnNameList, insertTargetEntry->resname);
+	}
+
+	/* set up a DestReceiver that copies into the distributed table */
+	copyDest = CreateCitusCopyDestReceiver(targetRelationId, columnNameList,
+										   executorState, stopOnFailure);
+
+	ExecuteIntoDestReceiver(selectQuery, paramListInfo, (DestReceiver *) copyDest);
+
+	executorState->es_processed = copyDest->tuplesSent;
+
+	XactModificationLevel = XACT_MODIFICATION_DATA;
+}
+
+
+/*
+ * ExecuteIntoDestReceiver plans and executes a query and sends results to the given
+ * DestReceiver.
+ */
+static void
+ExecuteIntoDestReceiver(Query *query, ParamListInfo params, DestReceiver *dest)
+{
+	PlannedStmt *queryPlan = NULL;
+	Portal portal = NULL;
+	int eflags = 0;
+	int cursorOptions = 0;
+	long count = FETCH_ALL;
+
+	/* create a new portal for executing the query */
+	portal = CreateNewPortal();
+
+	/* don't display the portal in pg_cursors, it is for internal use only */
+	portal->visible = false;
+
+#if (PG_VERSION_NUM >= 90600)
+	cursorOptions = CURSOR_OPT_PARALLEL_OK;
+#endif
+
+	/* plan the subquery, this may be another distributed query */
+	queryPlan = pg_plan_query(query, cursorOptions, params);
+
+	PortalDefineQuery(portal,
```

Comment:
I've got a question and a note here. 
It seems when we use `Portal`, it doesn't create a separate session for the `subquery`:

```SQL
SET citus.shard_replication_factor = 1;
SET citus.shard_count = 4;


CREATE TABLE events_table (user_id int, time timestamp, event_type int, value_2 int, value_3 float, value_4 bigint);
SELECT create_distributed_table('events_table', 'user_id');
COPY events_table_local FROM '/Users/onderkalaci/Documents/citus_code/citus/src/test/regress/data/events_table.data' WITH CSV;


CREATE TABLE agg_results (user_id int, value_1_agg int, value_2_agg int, value_3_agg float, value_4_agg bigint, agg_time timestamp);
SELECT create_distributed_table('agg_results', 'user_id');

INSERT INTO agg_results (value_4_agg, user_id) SELECT user_id, value_3 FROM events_table WHERE user_id = 5;
DEBUG:  StartTransactionCommand
DEBUG:  StartTransaction
DEBUG:  name: unnamed; blockState:       DEFAULT; state: INPROGR, xid/subid/cid: 0/1/0, nestlvl: 1, children: 
DEBUG:  cannot perform distributed INSERT INTO ... SELECT because the partition columns in the source table and subquery do not match
DETAIL:  The data type of the target table's partition column should exactly match the data type of the corresponding simple column reference in the subquery.
DEBUG:  ProcessQuery
DEBUG:  Collecting INSERT ... SELECT results on coordinator
DEBUG:  Creating router plan
DEBUG:  Plan is router executable
ERROR:  cannot establish new placement connection for DML when other connections exist for colocated placements

```
Do you foresee any other such cases? 

(As far as I can see, `SPI` wouldn't be applicable here. )